core/server/common/src/main/java/alluxio/master/journal/JournalTool.java

```diff
@@ -239,4 +194,299 @@ private static void usage() {
+              break;
+            case LOG:
+              JournalEntry entry = reader.getEntry();
+              out.println(ENTRY_SEPARATOR);
+              out.print(entry);
+              break;
+            case DONE:
+              done = true;
+              break;
+            default:
+              throw new RuntimeException("Unknown state: " + state);
+          }
+        }
+      }
+    }
+
+    /**
+     * @return the journal location
+     */
+    private URI getJournalLocation(String inputDir) {
+      if (!inputDir.endsWith(AlluxioURI.SEPARATOR)) {
+        inputDir += AlluxioURI.SEPARATOR;
+      }
+      try {
+        return new URI(inputDir);
+      } catch (URISyntaxException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  static class RaftJournalDumper extends AbstractJournalDumper {
+
+    /**
+     * Creates embedded journal dumper.
+     *
+     * @param master journal master
+     * @param start journal start sequence
+     * @param end journal end sequence
+     * @param outputDir output dir for journal dump
+     * @param inputDir input dir for journal files
+     */
+    public RaftJournalDumper(String master, long start, long end, String outputDir, String inputDir)
+        throws IOException {
+      super(master, start, end, outputDir, inputDir);
+    }
+
+    @Override
+    void dumpJournal() throws Throwable {
+      readFromDir();
+    }
+
+    /**
+     * Reads from the journal directly instead of going through the raft cluster. The state read
+     * this way may be stale, but it can still be useful for debugging while the cluster is offline.
+     */
+    private void readFromDir() throws Throwable {
+      Serializer serializer = RaftJournalSystem.createSerializer();
+      serializer.resolve(new ClientRequestTypeResolver());
+      serializer.resolve(new ClientResponseTypeResolver());
+      serializer.resolve(new ProtocolSerialization());
+      serializer.resolve(new ServerSerialization());
+      serializer.resolve(new StorageSerialization());
+
+      SingleThreadContext context = new SingleThreadContext("readJournal", serializer);
+
+      try {
+        // Read through the whole journal content, starting from snapshot.
+        context.execute(this::readCopycatSnapshotFromDir).get();
+        context.execute(this::readCopycatLogFromDir).get();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw e;
+      } catch (ExecutionException e) {
+        throw e.getCause();
+      }
+    }
+
+    private void readCopycatLogFromDir() {
+      try (PrintStream out =
+          new PrintStream(new BufferedOutputStream(new FileOutputStream(mJournalEntryFile)))) {
+        Log log = Storage.builder().withDirectory(mInputDir).build().openLog("copycat");
+        for (long i = log.firstIndex(); i < log.lastIndex(); i++) {
+          io.atomix.copycat.server.storage.entry.Entry entry = log.get(i);
+          if (entry instanceof CommandEntry) {
```

Comment:
Are command entries the only ones we care about and are all command entries our journal entries that we want to output?

Reply:
`writeSelected()` does the filtering for edit logs.
For snapshots though we can't do filtering. (Or at least it's not easy)